src/protocols/secure_channel/PASESession.cpp

```diff
@@ -578,115 +651,131 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
 
     if (err != CHIP_NO_ERROR)
     {
-        SendErrorMsg(Spake2pErrorType::kUnexpected);
+        SendStatusReport(kProtocolCodeInvalidParam);
     }
     return err;
 }
 
-CHIP_ERROR PASESession::HandleMsg2_and_SendMsg3(const System::PacketBufferHandle & msg)
+CHIP_ERROR PASESession::HandleMsg2_and_SendMsg3(System::PacketBufferHandle && msg)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
     uint8_t verifier[kMAX_Hash_Length];
-    size_t verifier_len_raw = kMAX_Hash_Length;
-    uint16_t verifier_len; // To be inited one we check length is small enough
-
-    uint8_t * buf  = msg->Start();
-    size_t buf_len = msg->DataLength();
+    size_t verifier_len = kMAX_Hash_Length;
 
     System::PacketBufferHandle resp;
 
-    Spake2pErrorType spake2pErr = Spake2pErrorType::kUnexpected;
-
-    uint16_t encryptionKeyId = 0;
+    uint16_t spake2pErr = kProtocolCodeInvalidParam;
 
     ChipLogDetail(SecureChannel, "Received spake2p msg2");
 
-    VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
-    VerifyOrExit(buf_len == sizeof(encryptionKeyId) + kMAX_Point_Length + kMAX_Hash_Length,
-                 err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
 
-    encryptionKeyId = chip::Encoding::LittleEndian::Read16(buf);
-    msg->ConsumeHead(sizeof(encryptionKeyId));
-    buf     = msg->Start();
-    buf_len = msg->DataLength();
+    const uint8_t * Y;
+    size_t Y_len = 0;
 
-    ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", encryptionKeyId);
-    SetPeerKeyId(encryptionKeyId);
+    const uint8_t * peer_verifier;
+    size_t peer_verifier_len = 0;
 
-    err = mSpake2p.ComputeRoundTwo(buf, kMAX_Point_Length, verifier, &verifier_len_raw);
-    SuccessOrExit(err);
-    VerifyOrExit(CanCastTo<uint16_t>(verifier_len_raw), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
-    verifier_len = static_cast<uint16_t>(verifier_len_raw);
+    uint32_t decodeTagIdSeq = 0;
 
-    {
-        const uint8_t * hash = &buf[kMAX_Point_Length];
-        err                  = mSpake2p.KeyConfirm(hash, kMAX_Hash_Length);
-        if (err != CHIP_NO_ERROR)
-        {
-            spake2pErr = Spake2pErrorType::kInvalidKeyConfirmation;
-            SuccessOrExit(err);
-        }
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-        err = mSpake2p.GetKeys(mKe, &mKeLen);
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    Y_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(Y));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    peer_verifier_len = tlvReader.GetLength();
```

Comment:
Are we certain all the Spake2p module methods will return error if any input is of the wrong size? If not, we should validate that incoming byte strings match expected lengths prior to passing them down.

Reply:
Yes, Spake2p code is checking for the lengths of incoming buffers. It returns error if the check fails.